src/utils.rs

```diff
@@ -158,40 +160,44 @@ pub(crate) fn random_plusminus_by_size_with_minimum<R: RngCore + CryptoRng>(
 /// [`Transcript`].
 pub(crate) fn plusminus_bn_random_from_transcript(
     transcript: &mut Transcript,
-    n: &BigNumber,
-) -> BigNumber {
+    _n: &BigNumber,
+) -> Result<BigNumber> {
     let mut is_neg_byte = vec![0u8; 1];
     transcript.challenge_bytes(b"sampling negation bit", is_neg_byte.as_mut_slice());
     let is_neg: bool = is_neg_byte[0] & 1 == 1;
 
     // The sampling method samples from the open interval, so add 1 to sample from
     // the _closed_ interval we want here.
-    let open_interval_max = n + 1;
-    let b = positive_bn_random_from_transcript(transcript, &open_interval_max);
+    let q = k256_order();
+    let open_interval_max = &q + 1;
+    let b = positive_bn_random_from_transcript(transcript, &open_interval_max)?;
     match is_neg {
-        true => -b,
-        false => b,
+        true => Ok(-b),
+        false => Ok(b),
     }
 }
 
-/// Derive a deterministic pseduorandom value in `[0, n)` from the
-/// [`Transcript`].
 pub(crate) fn positive_bn_random_from_transcript(
     transcript: &mut Transcript,
-    n: &BigNumber,
-) -> BigNumber {
-    let len = n.to_bytes().len();
-    let mut t = vec![0u8; len];
+    _n: &BigNumber,
+) -> Result<BigNumber> {
     // To avoid sample bias, we can't take `t mod n`, because that would bias
     // smaller numbers. Instead, we re-sample a new value (different because
     // there's a new label in the transcript).
-    loop {
+    let q = k256_order();
+    let len = q.to_bytes().len();
```

Comment:
todo: I thought based on the the way this function is used, we can't use `q` as the bound for every scenario. What are the actual expected bounds in the places where this is called? If it's not `q` in every place, we have to use the bound passed into the function here, not `q`. You might need to cross-reference with the paper to answer this.

The tests might pass if we're using `q` as the bound because `q` is smaller than the moduli `N` that we use, so something sampled from the natural numbers < `q` will by definition be in the natural numbers < `N`, but we still need to make sure we're doing what the paper requires. The tests don't ensure correctness against the paper.

Reply:
In the paper for Pi_mod, it is N=pq. So shall I change it back to n passed to the function? And remove q,ie, k256_order altogether?

Reply:
Yes, that sounds correct to me.